src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c

```diff
@@ -2072,6 +2198,94 @@ static ssize_t vmr_debug_type_store(struct device *dev,
 }
 static DEVICE_ATTR_WO(vmr_debug_type);
 
+static ssize_t clk_scaling_stat_raw_show(struct device *dev,
+	struct device_attribute *attr, char *buf)
+{
+	struct xocl_xgq_vmr *xgq = platform_get_drvdata(to_platform_device(dev));
+	ssize_t cnt = 0;
+
+	//Read clock scaling configuration settings
+	cnt += sprintf(buf + cnt, "HAS_CLOCK_THROTTLING:%d\n", xgq->xgq_cs_struct.has_clk_scaling);
+	cnt += sprintf(buf + cnt, "CLOCK_THROTTLING_ENABLED:%d\n", xgq->xgq_cs_struct.clk_scaling_en);
+	cnt += sprintf(buf + cnt, "POWER_SHUTDOWN_LIMIT:%u\n", xgq->xgq_cs_struct.pwr_shutdown_limit);
+	cnt += sprintf(buf + cnt, "TEMP_SHUTDOWN_LIMIT:%u\n", xgq->xgq_cs_struct.temp_shutdown_limit);
+	cnt += sprintf(buf + cnt, "POWER_THROTTLING_LIMIT:%u\n", xgq->xgq_cs_struct.pwr_scaling_limit);
+	cnt += sprintf(buf + cnt, "TEMP_THROTTLING_LIMIT:%u\n", xgq->xgq_cs_struct.temp_scaling_limit);
+	cnt += sprintf(buf + cnt, "POWER_THROTTLING_OVRD_LIMIT:%u\n", xgq->xgq_cs_struct.pwr_scaling_ovrd_limit);
+	cnt += sprintf(buf + cnt, "TEMP_THROTTLING_OVRD_LIMIT:%u\n", xgq->xgq_cs_struct.temp_scaling_ovrd_limit);
+	cnt += sprintf(buf + cnt, "POWER_THROTTLING_OVRD_ENABLE:%u\n", xgq->xgq_cs_struct.pwr_scaling_ovrd_limit);
+	cnt += sprintf(buf + cnt, "TEMP_THROTTLING_OVRD_ENABLE:%u\n", xgq->xgq_cs_struct.temp_scaling_ovrd_limit);
+	cnt += sprintf(buf + cnt, "CLOCK_THROTTLING_MODE:%u\n", xgq->xgq_cs_struct.clk_scaling_mode);
+
+	return cnt;
+}
+static DEVICE_ATTR_RO(clk_scaling_stat_raw);
+
+static ssize_t clk_scaling_configure_store(struct device *dev,
+                                           struct device_attribute *attr,
+                                           const char *buf, size_t count)
+{
+	struct xocl_xgq_vmr *xgq = platform_get_drvdata(to_platform_device(dev));
+	uint8_t enable = 0;
+	uint16_t pwr = 0;
+	uint8_t temp = 0;
+	char* pch = buf;
+	char* end = buf;
+	int ret = 0;
+
+	if (!xgq->xgq_cs_struct.has_clk_scaling)
+	{
+		XGQ_ERR(xgq, "clock scaling feature is not supported");
+		return -ENOTSUPP;
+	}
+
+	if (pch != NULL) {
+		pch = strsep(&end, ",");
+		if (kstrtou8(pch, 10, &enable) == -EINVAL || enable > 1) {
+			XGQ_ERR(xgq, "value should be 0 (disable) or 1 (enable)");
+			return -EINVAL;
+		}
+		pch = end;
+	}
+
+	if (pch != NULL) {
+		pch = strsep(&end, ",");
+		if ((kstrtou16(pch, 10, &pwr) == -EINVAL) ||
+			(pwr > xgq->xgq_cs_struct.pwr_scaling_limit)) {
+			XGQ_ERR(xgq, "Invalid power override limit %u provided, whereas max limit is %u",
+					pwr, xgq->xgq_cs_struct.pwr_scaling_limit);
+			return -EINVAL;
+		}
+		pch = end;
+	}
+
+	if (pch != NULL) {
+		pch = strsep(&end, ",");
+		if ((kstrtou8(pch, 10, &temp) == -EINVAL) ||
+			(temp > xgq->xgq_cs_struct.temp_scaling_limit)) {
+			XGQ_ERR(xgq, "Invalid temp override limit %u provided, wereas max limit is %u",
+					temp, xgq->xgq_cs_struct.temp_scaling_limit);
+			return -EINVAL;
+		}
+		pch = end;
+	}
+	mutex_lock(&xgq->xgq_lock);
```

Comment:
This is very good, maybe the vmr_status also need mutex to protect. Otherwise, the return payload might be overrided to corrupted data. I will fix my code.